File: slideholder/src/main/java/zhan/library/slide/holder/SlideViewHolder.java
package zhan.library.slide.holder;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.view.View;

import androidx.recyclerview.widget.RecyclerView;

import zhan.library.slide.ISlide;
import zhan.library.slide.SlideAnimatorListener;
import zhan.library.slide.SlideAnimatorUpdateListener;
import zhan.library.slide.helper.SlideAnimationHelper;


public abstract class SlideViewHolder extends RecyclerView.ViewHolder implements ISlide {

    private static final int DURATION_OPEN = 300;
    private static final int DURATION_CLOSE = 150;
    private static final int NORMAL_OFFSET = 50;

    private SlideAnimationHelper mSlideAnimationHelper = new SlideAnimationHelper();

    private SlideAnimatorListener slideAnimatorListener;
    private SlideAnimatorUpdateListener slideAnimatorUpdateListener;

    protected int mOffset;
    private int openDuration = DURATION_OPEN;
    private int closeDuration = DURATION_CLOSE;

    public SlideViewHolder(View itemView) {
        super(itemView);
        mOffset = SlideAnimationHelper.getOffset(itemView.getContext(), NORMAL_OFFSET);
        mSlideAnimationHelper.addAnimatorListener(new InAnimatorListener());
        mSlideAnimationHelper.addAnimatorUpdateListener(new InUpdateListener());
    }

    public void setSlideAnimatorListener(SlideAnimatorListener listener) {
        this.slideAnimatorListener = listener;
    }

    public void setSlideAnimatorUpdateListener(SlideAnimatorUpdateListener listener) {
        this.slideAnimatorUpdateListener = listener;
    }

    public void setOffset(int offset) {
        mOffset = SlideAnimationHelper.getOffset(itemView.getContext(), offset);
    }

    public int getOffset() {
        return mOffset;
    }

    public int getState() {
        return mSlideAnimationHelper.getState();
    }

    public SlideAnimationHelper getmSlideAnimationHelper() {
        return mSlideAnimationHelper;
    }

    public int getOpenDuration() {
        return openDuration;
    }

    public void setOpenDuration(int openDuration) {
        this.openDuration = openDuration;
    }

    public int getCloseDuration() {
        return closeDuration;
    }

    public void setCloseDuration(int closeDuration) {
        this.closeDuration = closeDuration;
    }

    //keep change state
    public void onBindSlide(View targetView) {
        switch (mSlideAnimationHelper.getState()) {
            case SlideAnimationHelper.STATE_CLOSE:
                targetView.scrollTo(0, 0);
                onBindSlideClose(SlideAnimationHelper.STATE_CLOSE);
                break;

            case SlideAnimationHelper.STATE_OPEN:
                targetView.scrollTo(-mOffset, 0);
                doAnimationSetOpen(SlideAnimationHelper.STATE_OPEN);
                break;
        }
    }

    @Override
    public void slideOpen() {
        mSlideAnimationHelper.openAnimation(openDuration);
    }

    @Override
    public void slideClose() {
        mSlideAnimationHelper.closeAnimation(closeDuration);
    }

    public void doAnimationSet(int offset, float fraction) {
        //empty
    }

    public void onBindSlideClose(int state) {
        //empty
    }

    public void doAnimationSetOpen(int state) {
        //empty
    }

    public void cancel() {
        mSlideAnimationHelper.cancel();
    }

    protected void handlerOpenAnimationUpdate(ValueAnimator animation) {
        float fraction = animation.getAnimatedFraction();
        int endX = (int) (-mOffset * fraction);
        doAnimationSet(endX, fraction);
    }

    protected void handlerCloseAnimationUpdate(ValueAnimator animation) {
        float fraction = animation.getAnimatedFraction();
        fraction = 1.0f - fraction;
        int endX = (int) (-mOffset * fraction);
        doAnimationSet(endX, fraction);
    }

    private class InUpdateListener implements ValueAnimator.AnimatorUpdateListener {

        @Override
        public void onAnimationUpdate(ValueAnimator animation) {
            if (mSlideAnimationHelper.isOpenStatus()) {
                handlerOpenAnimationUpdate(animation);

            } else {
                handlerCloseAnimationUpdate(animation);
            }

            if (slideAnimatorUpdateListener != null) {
                slideAnimatorUpdateListener.onAnimationUpdate(animation);
                slideAnimatorUpdateListener.onSlideAnimationUpdate(animation,
                        mSlideAnimationHelper.getState());
            }
        }
    }

    private class InAnimatorListener implements Animator.AnimatorListener {

        @Override
        public void onAnimationStart(Animator animation) {
            if (slideAnimatorListener != null) {
                slideAnimatorListener.onAnimationStart(animation);
                slideAnimatorListener.onSlideAnimationStart(animation,
                        mSlideAnimationHelper.getState());
            }
        }

        @Override
        public void onAnimationEnd(Animator animation) {
            if (slideAnimatorListener != null) {
                slideAnimatorListener.onAnimationEnd(animation);
                slideAnimatorListener.onSlideAnimationEnd(animation,
                        mSlideAnimationHelper.getState());
            }
        }

        @Override
        public void onAnimationCancel(Animator animation) {
            if (slideAnimatorListener != null) {
                slideAnimatorListener.onAnimationCancel(animation);
            }
        }

        @Override
        public void onAnimationRepeat(Animator animation) {
            if (slideAnimatorListener != null) {
                slideAnimatorListener.onAnimationRepeat(animation);
            }
        }
    }
}

File: slideholder/src/main/java/zhan/library/slide/holder/SlideViewHolder.kt
package zhan.library.slide.holder

import android.animation.Animator
import android.animation.ValueAnimator
import android.view.View

import androidx.recyclerview.widget.RecyclerView

import zhan.library.slide.ISlide
import zhan.library.slide.SlideAnimatorListener
import zhan.library.slide.SlideAnimatorUpdateListener
import zhan.library.slide.helper.SlideAnimationHelper


abstract class SlideViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), ISlide {

    private val mSlideAnimationHelper = SlideAnimationHelper()

    private var slideAnimatorListener: SlideAnimatorListener? = null
    private var slideAnimatorUpdateListener: SlideAnimatorUpdateListener? = null

    protected var mOffset: Int = 0
    private var openDuration = DURATION_OPEN
    private var closeDuration = DURATION_CLOSE

    var offset: Int
        get() = mOffset
        set(offset) {
            mOffset = SlideAnimationHelper.getOffset(itemView.context, offset)
        }

    val state: Int
        get() = mSlideAnimationHelper.state

    init {
        mOffset = SlideAnimationHelper.getOffset(itemView.context, NORMAL_OFFSET)
        mSlideAnimationHelper.addAnimatorListener(InAnimatorListener())
        mSlideAnimationHelper.addAnimatorUpdateListener(InUpdateListener())
    }

    fun setSlideAnimatorListener(listener: SlideAnimatorListener) {
        this.slideAnimatorListener = listener
    }

    fun setSlideAnimatorUpdateListener(listener: SlideAnimatorUpdateListener) {
        this.slideAnimatorUpdateListener = listener
    }

    fun getmSlideAnimationHelper(): SlideAnimationHelper {
        return mSlideAnimationHelper
    }

    //keep change state
    fun onBindSlide(targetView: View) {
        when (mSlideAnimationHelper.state) {
            SlideAnimationHelper.STATE_CLOSE -> {
                targetView.scrollTo(0, 0)
                onBindSlideClose(SlideAnimationHelper.STATE_CLOSE)
            }

            SlideAnimationHelper.STATE_OPEN -> {
                targetView.scrollTo(-mOffset, 0)
                doAnimationSetOpen(SlideAnimationHelper.STATE_OPEN)
            }
        }
    }

    override fun slideOpen() {
        mSlideAnimationHelper.openAnimation(openDuration.toLong())
    }

    override fun slideClose() {
        mSlideAnimationHelper.closeAnimation(closeDuration.toLong())
    }

    open fun doAnimationSet(offset: Int, fraction: Float) {
        //empty
    }

    open fun onBindSlideClose(state: Int) {
        //empty
    }

    open fun doAnimationSetOpen(state: Int) {
        //empty
    }

    fun cancel() {
        mSlideAnimationHelper.cancel()
    }

    protected fun handlerOpenAnimationUpdate(animation: ValueAnimator) {
        val fraction = animation.animatedFraction
        val endX = (-mOffset * fraction).toInt()
        doAnimationSet(endX, fraction)
    }

    protected fun handlerCloseAnimationUpdate(animation: ValueAnimator) {
        var fraction = animation.animatedFraction
        fraction = 1.0f - fraction
        val endX = (-mOffset * fraction).toInt()
        doAnimationSet(endX, fraction)
    }

    private inner class InUpdateListener : ValueAnimator.AnimatorUpdateListener {

        override fun onAnimationUpdate(animation: ValueAnimator) {
            if (mSlideAnimationHelper.isOpenStatus) {
                handlerOpenAnimationUpdate(animation)

            } else {
                handlerCloseAnimationUpdate(animation)
            }

            if (slideAnimatorUpdateListener != null) {
                slideAnimatorUpdateListener!!.onAnimationUpdate(animation)
                slideAnimatorUpdateListener!!.onSlideAnimationUpdate(animation,
                        mSlideAnimationHelper.state)
            }
        }
    }

    private inner class InAnimatorListener : Animator.AnimatorListener {

        override fun onAnimationStart(animation: Animator) {
            if (slideAnimatorListener != null) {
                slideAnimatorListener!!.onAnimationStart(animation)
                slideAnimatorListener!!.onSlideAnimationStart(animation,
                        mSlideAnimationHelper.state)
            }
        }

        override fun onAnimationEnd(animation: Animator) {
            if (slideAnimatorListener != null) {
                slideAnimatorListener!!.onAnimationEnd(animation)
                slideAnimatorListener!!.onSlideAnimationEnd(animation,
                        mSlideAnimationHelper.state)
            }
        }

        override fun onAnimationCancel(animation: Animator) {
            if (slideAnimatorListener != null) {
                slideAnimatorListener!!.onAnimationCancel(animation)
            }
        }

        override fun onAnimationRepeat(animation: Animator) {
            if (slideAnimatorListener != null) {
                slideAnimatorListener!!.onAnimationRepeat(animation)
            }
        }
    }

    companion object {

        private const val DURATION_OPEN = 300
        private const val DURATION_CLOSE = 150
        private const val NORMAL_OFFSET = 50
    }
}
